README.md
# @auroradao/datastream-client

This monorepo provides the necessary packages to easily connect your Javascript/Typescript applications to the IDEX Datastream API. It includes various connectors allowing you to choose the best `WebSocket` implementation based upon your environment.

The client has been built to be as simple as possible to provide a rock-solid connection. We already use the web implementation for our web client and the [`uws`](https://github.com/uNetworking/uWebSockets) for any backend integrations we require.

## Features

- **Web & Node** - No matter your environment we have you covered! We maintain connectors for web and multiple node WebSocket implementations!
- **TypeScript** - All packages are written with [`TypeScript`](https://www.typescriptlang.org) giving you type-safety and awesome IDE features while still fully compatible for Javascripters out there.
- **Datastream Handshake** - The handshake process is handled for you, just create the client and start receiving events!
- **Automatic Reconnect** - Should the connection to the API terminate, it will be maintained and reconnected to for you automatically.
  - We implemented the `exponential-backoff` required by the Datastream specification so you don't have to!
- **Offline Buffer** - No need to worry about connection states, your requests will be sent automatically if possible once the connection is re-established.
- **Stateful Connections** - The Datastream is a session-based connection. As-such, any disconnection means you will have to re-subscribe to your desired events. By turning on the `stateful` feature, the client will attempt to maintain this state for you.

## Installation (Web)

Nothing special is required when using the client on the web, we use the `@auroradao/datastream-connector-web` connector by default so you don't need to provide one!

```bash
yarn add @auroradao/datastream-client
```

## Installation (Node)

On Node we need to provide a `WebSocket` "connector" that we wish to use. A connector is a simple interface that normalizes a given implementation to provide us with the values we expect. You can even provide your own easily!

We currently maintain two packages:

- [`@auroradao/datastream-connector-ws`](./packages/datastream-connector-ws) using the [`ws`](https://github.com/websockets/ws) package
- [`@auroradao/datastream-connector-uws`](./packages/datastream-connector-uws) using an implementation of the [`uws`](https://github.com/uNetworking/uWebSockets) binding for Node.

```bash
yarn add @auroradao/datastream-client
yarn add @auroradao/datastream-connector-uws
```

## Example (Web)

```javascript
import createDatastreamClient from '@auroradao/datastream-client';

const client = createDatastreamClient(
  {
    log: true,
    key: '<API-KEY>',
  },
  {
    onConnect() {
      client.subscribe('account', '0x..', [
        'account_withdrawal_dispatched',
        'account_withdrawal_complete',
      ]);
      // any request may also be used as a promise
      // with a timeout if we need to wait for
      // the result
      client
        .subscribe(
          'markets',
          ['ETH_IDEX', 'ETH_IDXM'],
          ['market_orders', 'market_cancels', 'market_trades'],
        )
        .promise({ timeout: 10000 })
        .then(result => {
          // subscription successful!
        })
        .catch(err => {
          if (err.message === 'TIMEOUT') {
            // subscription timed out!
          } else {
            // subscription failed!
          }
        });
    },
    onEvent(message) {
      console.log('Event: ', message);
      switch (message.event) {
        case 'market_trades':
        case 'market_cancels':
        case 'market_orders': {
          break;
        }
        case 'account_withdrawal_dispatched':
        case 'account_withdrawal_complete': {
          break;
        }
        default: {
          console.log('Unknown Event: ', message.event);
          break;
        }
      }
    },
  },
);
```

## Example (Node)

```javascript
import createDatastreamClient from '@auroradao/datastream-client';
import uwsConnector from '@auroradao/datastream-connector-uws';

const client = createDatastreamClient({
  log: true,
  key: '<API-KEY>',
  connector: uwsConnector,
});
```

## API Reference

### Initial Configuration

```javascript
interface BufferConfiguration {
  size: number;
}

interface InitialConfiguration {
  /**
   * The API Key that you wish to use to authenticate
   * with the Datastream server.
   *
   * @required
   */
  key: string;
  /**
   * If using a connector other than HTML5, provide
   * the connector in the initial configuration
   * here.
   */
  connector?: Connection$Connector;
  /**
   * Should the client automatically maintain a persistent
   * connection to the server?
   *
   * @note
   *  This will automatically begin connecting immediately
   *  after the client is created.
   *
   * @defaultValue true
   */
  auto?: boolean;
  /**
   * Optionally provide the type of client connection
   * that should be established.
   *
   * @note
   *  In most cases this should not need to be changed
   *  from the default value.
   *
   * @defaultValue "client"
   */
  type?: string;
  /**
   * Enable or disable the clients internal logging.
   *
   * @defaultValue false
   */
  log?: boolean;
  /**
   * Optionally provide a WebSocket URL to use instead
   * of the default server.
   *
   * @defaultValue "wss://datastream.idex.market"
   */
  url?: string;
  /**
   * A stateful client will attempt to automatically
   * restore its subscriptions and general state if
   * it needs to reconnect at any point.
   *
   * @defaultValue false
   */
  stateful?: boolean;
  /**
   * You may optionally provide a locale which will
   * indicate what language should be returned for
   * any messages or errors.
   *
   * @note
   *  If a translation is not available for a given
   *  message or error, it may instead return "en"
   *  (English).
   *
   * @defaultValue "en"
   */
  locale?: string;
  buffer?: BufferConfiguration;
}
```

### Subscriptions

Below is the general information about the available Datastream `subscriptions`, including example payloads. \

> **Important:** Values given in the example payloads are not precise examples and may be abbreviated or changed. They are only meant to represent the general form the event will take.

> **Reminder:** `0x0000000000000000000000000000000000000000` represents `ETH` whenever it is provided for a given address.

#### `client.subscribe()`

Subscribing to events is done by calling the `subscribe` method on a given client. This function has the signature as shown below. For the complete set of type signatures you can view them in the [Datastream Types Package](./packages/datastream-types).

```javascript
type Subscribe$Categories =
  | 'accounts'
  | 'markets'
  | 'chains'

subscribe(
  to: Subscribe$Categories,
  topics: string | string[],
  events: string | string[]
): Client$SendResponse<string, Subscribe$Requests>;
```

A client may subscribe to many events for a given topic. However, each request must contain the entirety of the subscription for a given topic. For example, subscribing to multiple markets would look like this:

```javascript
client.subscribe(
  'markets',
  ['ETH_AURA', 'ETH_IDXM'],
  ['market_orders', 'market_cancels', 'market_trades'],
);
```

---

#### Account Events

##### account_nonce

Received whenever the subscribed accounts `nonce` has been updated. This new nonce must be used as the base nonce for all future requests.

```javascript
{
  sid: 'sid:ehRMUyHAc26h',
  eid: 'evt:D1_ja4kmeeDc',
  event: 'account_none',
  seq: 2,
  payload: {
    account: '0x...',
    nonce: '999'
  },
}
```

> When using Javascript, it is safest to use a `BigNumber` library (or [`BigInt`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt) if your environment supports it). Below is example code of incrementing the `nonce` value by 1 when needed using [`BigInt`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt):

```javascript
// nice and simple :-)
const currentNonce = BigInt(event.payload.nonce);
const nextNonce = currentNonce + 1n;
```

##### account_deposit_complete

Received when a `deposit` is received and credited to the account. At this point the deposited funds are available for trading.

```javascript
{

  sid: 'sid:PKr79dtJs5T',
  eid: 'evt:zIABS7OTq',
  event: 'account_deposit_complete',
  seq: 84,
  payload: {
    account: '0x...',
    deposit: {
      id: 348738,
      user: '0x...',
      token: '0x0000000000000000000000000000000000000000',
      amount: '1000000000000000000',
      transactionHash: '0x...',
      createdAt: '1969-01-01T01:01:01.000Z',
    },
  }
}
```

> **Note:** Deposits are not submitted using an API Request and are instead initiated directly with the IDEX Contract. This event is generated using our internal block scanner which is actively monitoring transactions on the network.

> **Note:** Deposit does not have `dispatched` and `pending` events associated with them as they are directly executed on the IDEX Client and are never processed by the IDEX Backend until they are seen in the blockchain.

##### account_orders

When the subscribed account has new `orders` received and processed by the exchange, this event will be provided including each of the `orders` that were processed in the given batch. At this point the orders should be considered `pending`.

```javascript
{
  sid: 'sid:ehRMUyHAc26h',
  eid: 'evt:U6_8jwkWFMe2',
  event: 'account_orders',
  seq: 9,
  payload: {
    account: '0x...',
    orders: [
      {
        id: 745729,
        amountBuy: '205795770000',
        amountSell: '1392477916542192014',
        tokenBuy: '0x3db6ba6ab6f95efed1a6e794cad492faaabf294d',
        tokenSell: '0x0000000000000000000000000000000000000000',
        nonce: 101010,
        hash: '0x...',
        user: '0x...',
        createdAt: '1969-01-01T01:01:01.000Z',
        updatedAt: '1969-01-01T01:01:01.000Z',
      },
    ],
  },
}
```

##### account_cancels

When the subscribed account has new `cancels` received and processed by the exchange, this event will be provided including each of the `cancels` that were processed in the given batch. At this point the `cancels` should be considered `pending`.

```javascript
{
  sid: 'sid:ehRMUyHAc26h',
  eid: 'evt:yiEXBY2UBe3e',
  event: 'account_cancels',
  seq: 10,
  payload: {
    account: '0x...',
    cancels: [
      {
        id: 987832,
        market: 'ETH_LTO',
        orderHash: '0x...',
        createdAt: '1969-01-01T01:01:01.000Z',
      },
    ],
  },
}
```

##### account_trades

When the subscribed account has new `trades` received and processed by the exchange, this event will be provided including each of the `trades` that were processed in the given batch. At this point the `trades` should be considered `pending`.

```javascript
{
  sid: 'sid:ehRMUyHAc26h',
  eid: 'evt:IVKYOfslj5e7',
  event: 'account_trades',
  seq: 172,
  payload: {
    account: '0x...',
    total: 1,
    highestTimestamp: 1552536625,
    trades: [
      {
        tid: 98989,
        type: 'buy',
        date: '1969-01-01T01:01:01.000Z',
        timestamp: 1552536625,
        market: 'ETH_LTO',
        usdValue: '411.171696117038254309',
        price: '0.000687600842795718',
        amount: '4557.68612841',
        total: '3.133868823093068736',
        taker: '0x...',
        maker: '0x...',
        orderHash: '0x...',
        gasFee: '1.483418775132356598',
        tokenBuy: '0x0000000000000000000000000000000000000000',
        tokenSell: '0x3db6ba6ab6f95efed1a6e794cad492faaabf294d',
        buyerFee: '4.55768612841',
        sellerFee: '0.002193708176165148',
        amountWei: '3133868823093068736',
        updatedAt: '1969-01-01T01:01:01.000Z',
      },
    ],
  },
};
```

##### account_withdrawal_created

A `withdrawal` request is first received by the server and queued to be dispatched to the blockchain. At this point the `withdrawal` should be considered as `pending`.

```javascript
{
  sid: 'sid:PKr79dtJs5T',
  eid: 'evt:hxp4-qwiYUFk',
  event: 'account_withdrawal_created',
  seq: 59,
  payload: {
    account: [
      '0x...',
      '0x...',
    ],
    withdrawal: {
      id: 909090,
      user: '0x...',
      amount: '1000000000000000000',
      token: '0x0000000000000000000000000000000000000000',
      nonce: 141,
      fee: '209999999800000',
      usdValue: '136.46',
      createdAt: '1969-01-01T01:01:01.000Z',
    },
  },
}
```

##### account_withdrawal_dispatched

A `withdrawal` request is first dispatched to the blockchain. At this point the `withdrawal` should be considered as `confirming`.

```javascript
{
  sid: 'sid:PKr79dtJs5T',
  eid: 'evt:6-VlUU1Wh',
  event: 'account_withdrawal_dispatched',
  seq: 62,
  payload: {
    account: ['0x...'],
    withdrawal: {
      sender: '0x...',
      id: 128,
      transactionHash: '0x...',
    },
  },
}
```

> **Tip:** At this point you will receive the associated `transactionHash` since it has been sent to the blockchain for processing.

##### account_withdrawal_complete

A `withdrawal` request is considered confirmed.

```javascript
{
  sid: 'sid:PKr79dtJs5T',
  eid: 'evt:HfW6bZhDZ',
  event: 'account_withdrawal_complete',
  seq: 74,
  payload: {
    withdrawal: {
      sender: '0x...',
      id: 128,
      transactionHash: '0x...',
      token: '0x0000000000000000000000000000000000000000',
      amount: '1000000000000000000',
    },
  },
}
```

##### account_trade_dispatched

A `trade` request is first dispatched to the blockchain. At this point the `trade` should be considered as `confirming`.

```javascript
{
  sid: 'sid:VtKO3C3fzJ5',
  eid: 'evt:ekkwGhwT3',
  event: 'account_trade_dispatched',
  seq: 11,
  payload: {
    account: [
      '0x...',
      '0x...'
    ],
    trade: {
      tid: 472,
      sender: '0x...',
      hash: '0x...',
      transactionHash: '0x...',
    },
  },
}
```

> **Tip:** In the payload, `hash` refers to the hash for the `order` while the `transactionHash` is the hash for the trade transaction which will appear on the blockchain.

##### account_trade_complete

A `trade` request is considered confirmed.

```javascript
{
  sid: 'sid:VtKO3C3fzJ5',
  eid: 'evt:Cgle_CpBh',
  event: 'account_trade_complete',
  seq: 25,
  payload: {
    account: [
      '0x...',
      '0x...',
    ],
    trade: {
      tid: 472,
      sender: '0x...',
      transactionHash: '0x...',
      hash: '0x...',
    },
  },
}
```

##### account_invalidation_dispatched

An `invalidation` request is first dispatched to the blockchain. At this point the `invalidation` should be considered as `confirming`.

##### account_invalidation_complete

An `invalidation` is considered confirmed.

##### account_balance_sheet

The updated balance sheet for the account. Triggered whenever the accounts balances are updated by a given action and includes all non-zero balances for the account. Balances are represented as [`WEI`](https://www.investopedia.com/terms/w/wei.asp).

```javascript
{
  sid: 'sid:PKr79dtJs5T',
  eid: 'evt:oMSLpNwq1fcR',
  event: 'account_balance_sheet',
  seq: 60,
  payload: {
    account: '0x...',
    balance: {
      '0x0000000000000000000000000000000000000000': '6656648607189529963',
      '0x...': '36517409284907215842',
    },
  },
}
```

> **Note:** Each value should be considered a `BigNumber` / `BigInt` value and should always be parsed as such. Using `Number` will almost certainly yield unwanted results.

##### account_rewards

```javascript
{
  sid: 'sid:VtKO3C3fzJ5',
  eid: 'evt:8tq0-D8NPwZH',
  event: 'account_rewards',
  seq: 7,
  payload: {
    account: '0x...',
    rewards: '602.930949384415100868',
  },
}
```

---

#### Market Events

##### market_orders

```javascript
{
  sid: 'sid:ehRMUyHAc26h',
  eid: 'evt:AB_9awkWFMe2',
  event: 'market_orders',
  seq: 229,
  payload: {
    market: 'ETH_LTO',
    orders: [
      {
        id: 101010,
        amountBuy: '205795770000',
        amountSell: '1392477916542192014',
        tokenBuy: '0x3db6ba6ab6f95efed1a6e794cad492faaabf294d',
        tokenSell: '0x0000000000000000000000000000000000000000',
        nonce: 101010,
        hash: '0x...',
        user: '0x...',
        createdAt: '1969-01-01T01:01:01.000Z',
        updatedAt: '1969-01-01T01:01:01.000Z',
      },
    ],
  },
}
```

##### market_cancels

```javascript
{
  sid: 'sid:ehRMUyHAc26h',
  eid: 'evt:xiAXRY2UAe2b',
  event: 'market_cancels',
  seq: 28,
  payload: {
    market: 'ETH_LTO',
    cancels: [
      {
        id: 101010,
        orderHash: '0x...',
        createdAt: '1969-01-01T01:01:01.000Z',
      },
    ],
  },
}
```

##### market_trades

```javascript
{
  sid: 'sid:ehRMUyHAc26h',
  eid: 'evt:ARvEax5lj5mm',
  event: 'market_trades',
  seq: 199,
  payload: {
    market: 'ETH_LTO',
    total: 1,
    highestTimestamp: 1552536625,
    trades: [
      {
        tid: 101010,
        type: 'buy',
        date: '1969-01-01T01:01:01.000Z',
        timestamp: 1552536625,
        market: 'ETH_LTO',
        usdValue: '411.171696117038254309',
        price: '0.000687600842795718',
        amount: '4557.68612841',
        total: '3.133868823093068736',
        taker: '0x...',
        maker: '0x...',
        orderHash: '0x...',
        gasFee: '1.483418775132356598',
        tokenBuy: '0x0000000000000000000000000000000000000000',
        tokenSell: '0x3db6ba6ab6f95efed1a6e794cad492faaabf294d',
        buyerFee: '4.55768612841',
        sellerFee: '0.002193708176165148',
        amountWei: '3133868823093068736',
        updatedAt: '1969-01-01T01:01:01.000Z',
      },
    ],
  },
};
```

---

#### Chain Events

##### chain_status

When the IDEX internal backend transitions to different states, such as temporarily disabled trading, this event will be dispatched.

```javascript
{
  sid: 'sid:EMKKI9kym1q',
  eid: 'evt:EbB21XwW16TP',
  event: 'chain_status',
  seq: 90,
  payload: {
    status: {
      restarting: false,
      trades: true,
      cancels: true,
      withdrawals: true,
    },
  },
}
```

##### chain_server_block

This event is dispatched to provide the current block that the IDEX Backend is currently processing. It does not necessarily indicate the current block that the network itself is on at any given dispatch.

```javascript
{
  sid: 'sid:ehRMUyHAc26h',
  eid: 'evt:JUZEAB0hu8tJ3',
  event: 'chain_server_block',
  seq: 622,
  payload: {
    block: 7364977
  },
}
```

##### chain_symbol_usd_price

Receives the latest usd pricing used by the exchange for the given symbol.

```javascript
{
  sid: 'sid:ehRMUyHAc26h',
  eid: 'evt:gKefqaoPKobEz',
  event: 'chain_symbol_usd_price',
  seq: 618,
  payload: {
    symbol: 'ETH/USD',
    price: 133.30
  },
}
```

##### chain_reward_pool_size

```javascript
{
  sid: 'sid:ehRMUyHAc26h',
  eid: 'evt:18BKijcJMd046',
  event: 'chain_reward_pool_size',
  seq: 682,
  payload: {
    pool: '12695555.461190489851413656'
  },
}
```

##### chain_gas_price

Provides the latest gas price in GWEI used by the exchange.

```javascript
{
  sid: 'sid:ehRMUyHAc26h',
  eid: 'evt:E9PRHmh7XFVws',
  event: 'chain_gas_price',
  seq: 361,
  payload: {
    price: '5'
  },
}
```

> **Note:** Please take note that the price is provided as a `string` here and not a number.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [1.9.2](https://github.com/AuroraDao/datastream-client-js/compare/v1.9.1...v1.9.2) (2019-12-03)

**Note:** Version bump only for package @auroradao/datastream-client-js

## [1.9.1](https://github.com/AuroraDao/datastream-client-js/compare/v1.9.0...v1.9.1) (2019-12-03)

### Bug Fixes

- **docs:** update docs to use ETH_IDEX instead of ETH_AURA ([b5fdd77](https://github.com/AuroraDao/datastream-client-js/commit/b5fdd77))

# [1.9.0](https://github.com/AuroraDao/datastream-client-js/compare/v1.8.8...v1.9.0) (2019-12-03)

### Features

- **datastream-client:** allow array of strings for topics ([2652c86](https://github.com/AuroraDao/datastream-client-js/commit/2652c86))

## [1.8.8](https://github.com/AuroraDao/datastream-client-js/compare/v1.8.7...v1.8.8) (2019-12-03)

### Bug Fixes

- **datastream-types:** fix message event types ([6317c8a](https://github.com/AuroraDao/datastream-client-js/commit/6317c8a))

## [1.8.7](https://github.com/AuroraDao/datastream-client-js/compare/v1.8.6...v1.8.7) (2019-12-03)

### Bug Fixes

- cosmiconfig dep ([878bf7a](https://github.com/AuroraDao/datastream-client-js/commit/878bf7a))
- resolutions for yarn ([4408caf](https://github.com/AuroraDao/datastream-client-js/commit/4408caf))
- set js-yaml to 3.13.1 ([3a5e316](https://github.com/AuroraDao/datastream-client-js/commit/3a5e316))
- set-value ^2.0.1 ([9c188ef](https://github.com/AuroraDao/datastream-client-js/commit/9c188ef))
- update other dev dependencies ([29547d8](https://github.com/AuroraDao/datastream-client-js/commit/29547d8))
- update set-value dependency ([c5985e4](https://github.com/AuroraDao/datastream-client-js/commit/c5985e4))

## [1.8.3](https://github.com/AuroraDao/datastream-client-js/compare/v1.8.2...v1.8.3) (2019-07-24)

### Bug Fixes

- check if connected before handshake ([01a1d0f](https://github.com/AuroraDao/datastream-client-js/commit/01a1d0f))

## [1.8.2](https://github.com/AuroraDao/datastream-client-js/compare/v1.8.1...v1.8.2) (2019-07-24)

### Bug Fixes

- fix for kyc token race condition ([8434aee](https://github.com/AuroraDao/datastream-client-js/commit/8434aee))

## [1.8.1](https://github.com/AuroraDao/datastream-client-js/compare/v1.8.0...v1.8.1) (2019-07-22)

### Bug Fixes

- fix configuration updates for token/key ([61d514c](https://github.com/AuroraDao/datastream-client-js/commit/61d514c))

# [1.8.0](https://github.com/AuroraDao/datastream-client-js/compare/v1.6.9...v1.8.0) (2019-07-22)

### Features

- add subscription to users ([#2](https://github.com/AuroraDao/datastream-client-js/issues/2)) ([3d6d495](https://github.com/AuroraDao/datastream-client-js/commit/3d6d495))
- resend handshake on key/token change ([b989fbd](https://github.com/AuroraDao/datastream-client-js/commit/b989fbd))

# [1.7.0](https://github.com/AuroraDao/datastream-client-js/compare/v1.6.9...v1.7.0) (2019-07-01)

### Bug Fixes

- **subscription:** add users to subscription message ([f108f2f](https://github.com/AuroraDao/datastream-client-js/commit/f108f2f))

### Features

- add subscription to users ([c1d63a0](https://github.com/AuroraDao/datastream-client-js/commit/c1d63a0))

## [1.6.9](https://github.com/AuroraDao/datastream-client-js/compare/v1.6.8...v1.6.9) (2019-04-18)

### Bug Fixes

- refactor of connectors to send socket value ([9032285](https://github.com/AuroraDao/datastream-client-js/commit/9032285))
- fix typing for web shims for reconnect ([5699f66](https://github.com/AuroraDao/datastream-client-js/commit/5699f66))
- only call socket.close in web if socket open ([2a0c998](https://github.com/AuroraDao/datastream-client-js/commit/2a0c998))
- more web connector shimminess ([1065d5d](https://github.com/AuroraDao/datastream-client-js/commit/1065d5d))
- attempt for better fix for browser reconnect ([5398769](https://github.com/AuroraDao/datastream-client-js/commit/5398769))
- handle reconnect() request so we dont schedule twice ([9b0ef0e](https://github.com/AuroraDao/datastream-client-js/commit/9b0ef0e))
- reconnect if connected or handshaked ([aae03d2](https://github.com/AuroraDao/datastream-client-js/commit/aae03d2))
- only reconnect if handshaked when connection.reconnect() is called ([ea5a54c](https://github.com/AuroraDao/datastream-client-js/commit/ea5a54c))
- use NORMAL for all closure codes ([2326ae3](https://github.com/AuroraDao/datastream-client-js/commit/2326ae3))

# [1.6.0](https://github.com/AuroraDao/datastream-client-js/compare/v1.5.0...v1.6.0) (2019-04-17)

### Features

- allow token config and setters for key/token ([94fe260](https://github.com/AuroraDao/datastream-client-js/commit/94fe260))
- upgrade to typescript 3.4 ([feee265](https://github.com/AuroraDao/datastream-client-js/commit/feee265))

# [1.5.0](https://github.com/AuroraDao/datastream-client-js/compare/v1.4.1...v1.5.0) (2019-04-08)

### Features

- update task-handler dependency ([76d38cd](https://github.com/AuroraDao/datastream-client-js/commit/76d38cd))

## [1.4.1](https://github.com/AuroraDao/datastream-client-js/compare/v1.4.0...v1.4.1) (2019-03-27)

### Bug Fixes

- fix case when error is standard style string ([30254de](https://github.com/AuroraDao/datastream-client-js/commit/30254de))

# [1.4.0](https://github.com/AuroraDao/datastream-client-js/compare/v1.3.2...v1.4.0) (2019-03-26)

### Features

- update task-handler and dev dependencies ([873d217](https://github.com/AuroraDao/datastream-client-js/commit/873d217))

## [1.3.2](https://github.com/AuroraDao/datastream-client-js/compare/v1.3.1...v1.3.2) (2019-03-25)

### Bug Fixes

- fixes for buffer of undefined error ([b572560](https://github.com/AuroraDao/datastream-client-js/commit/b572560))

# [1.3.0](https://github.com/AuroraDao/datastream-client-js/compare/v1.2.2...v1.3.0) (2019-03-20)

### Features


tests/quick.ts
/* This file is used to quickly test and iterate the "final package" by running it with
   ts-node-dev. When running `yarn test:quick`, it will run the file and watch for changes,
   automatically restarting when needed. */
import createDatastreamClient from '../packages/datastream-client';
// import wsConnector from '../packages/datastream-connector-ws';
import wsConnector from '../packages/datastream-connector-uws';

const client = createDatastreamClient(
  {
    log: true,
    url: 'wss://datastream.idex.market',
    key: '',
    connector: wsConnector,
  },
  {
    onConnect() {
      console.log('Connected!');
      this.subscribe('chains', 'eth', [
        'chain_symbol_usd_price',
        'chain_24hr_usd_volume',
      ]);
      this.subscribe('users', '24ccecd0-6385-4e64-8b6c-0dbf4c77c2f1', [
        'user_profile_updated',
      ]);
    },

    onMessage(message) {
      console.log('Message Received: ', message);
    },
  },
);


.eslintrc.js
module.exports = {
  root: true,
  parser: '@typescript-eslint/parser',
  parserOptions: {
    tsconfigRootDir: './packages',
    project: 'tsconfig.base.json',
  },
  extends: [
    'airbnb-base',
    'plugin:@typescript-eslint/recommended',
    'prettier',
    'prettier/@typescript-eslint',
  ],
  rules: {
    '@typescript-eslint/no-use-before-define': 'off',
    '@typescript-eslint/explicit-function-return-type': 'off',
    // cant handle Category$Name at the moment, although
    // pascal case should be enforced.
    '@typescript-eslint/class-name-casing': 'off',
    'comma-dangle': ['error', 'always-multiline'],
    'consistent-return': 'off',
    'no-restricted-syntax': 'off',
    'no-multi-assign': 'off',
    'no-use-before-define': 'off',
    'no-console': 'off',
    'no-underscore-dangle': 'off',
    // broken with tsc type aliases for now
    // waiting for my PR to merge before re-enable
    'import/named': 'warn',
    // typescript type imports suffer from this
    'import/no-cycle': 'off',
    'import/prefer-default-export': 'off',
    'import/no-extraneous-dependencies': [
      'error',
      { devDependencies: ['dev/**'] },
    ],
  },
  plugins: ['import', 'promise', 'prettier', '@typescript-eslint'],
  settings: {
    'import/extensions': ['.ts', '.tsx'],
    'import/parsers': {
      '@typescript-eslint/parser': ['.ts', '.tsx'],
    },
    'import/resolver': {
      typescript: {
        directory: './packages/tsconfig.base.json',
      },
    },
  },
};


wip/linked-list/src/node.ts
import { LinkedListNodeSymbol } from './constants';
import { ILinkedList, INode } from './types';

function removeNode<V>(l: ILinkedList<V>, node: INode<V>): INode<V> {
  if (!node.p && !node.n) {
    return node;
  }
  if (node.p) {
    node.p.n = node.n;
  } else {
    l.s = node.n;
  }
  if (node.n) {
    node.n.p = node.p;
  } else {
    l.e = node.p;
  }
  node.n = null;
  node.p = null;
  return node;
}

function insertNodeBefore<V>(
  l: ILinkedList<V>,
  self: INode<V>,
  inserted: INode<V>
): INode<V> {
  inserted.n = self;
  if (self.p === null) {
    self.p = inserted;
    inserted.p = null;
  } else {
    self.p.n = inserted;
    inserted.p = self.p;
    self.p = inserted;
  }
  if (inserted.p === null) {
    l.s = inserted;
  }
  return self;
}

function insertNodeAfter<V>(
  l: ILinkedList<V>,
  self: INode<V>,
  inserted: INode<V>
): INode<V> {
  inserted.p = self;
  if (self.n === null) {
    self.n = inserted;
  } else {
    self.n.p = inserted;
    inserted.n = self.n;
    inserted.p = self;
    self.n = inserted;
  }
  if (inserted.n === null) {
    l.e = inserted;
  }
  return self;
}

function isListNode<V>(v: V | INode<V>): v is INode<V> {
  return (
    typeof v === 'object' && (v as INode<V>)[LinkedListNodeSymbol] === true
  );
}

export default function ListNode<V>(
  l: ILinkedList<V>,
  v: V,
  p: null | INode<V> = null,
  n: null | INode<V> = null
): INode<V> {
  l.c += 1;
  const self: INode<V> = {
    [LinkedListNodeSymbol]: true,
    v,
    p,
    n,
    insert(pos: 'p' | 'n', next: V | INode<V>) {
      const inserted = isListNode(next) ? next : ListNode(l, next);
      if (inserted.n || inserted.p) {
        removeNode(l, inserted);
      }
      return pos === 'p'
        ? insertNodeBefore(l, self, inserted)
        : insertNodeAfter(l, self, inserted);
    },
    remove() {
      l.c -= 1;
      return removeNode(l, self);
    },
  };
  return self;
}


wip/linked-list/src/index.ts
import { ILinkedList } from './types';

import LinkedList from './list';

function start() {
  const list: ILinkedList<[string, number]> = LinkedList(
    new Map([['FIRST', 0]])
  );

  list.push(['key1', 1]);
  list.push(['key2', 2]);
  list.push(['key3', 3]);
  list.unshift(['key4', 4]);

  const list2 = LinkedList(list);
  list2.push(['NEW', 6]);

  list.forEach(v => console.log(v));
  list2.forEach(v => console.log(v));

  // let node = list.s;
  // while (node !== null) {
  //   const n = node.v[1];

  //   switch (n) {
  //     case 4:
  //     case 2:
  //     case 3: {
  //       node.insert('p', ['hi', list.c + 1]);
  //       break;
  //     }
  //   }
  //   node = node.n;
  // }
  // node = list.s;
  // // list.sort((a, b) => a[1] - b[1]);
  // // console.log(list.c);

  // // console.log(JSON.stringify(list));

  // list.insert(['new', 6], a => a[1] === 6);

  // const rlist = list.reverse;

  // rlist.insert(['new', 7], a => a[1] === 1);

  // console.log(JSON.stringify(list));
  // console.log(JSON.stringify(rlist));

  // console.log(JSON.stringify(list.merge(list)));
  // list.forEachReverse(v => console.log(v));
  // console.log('done');
  // const node = ListNode([1, 2]);
  // node.insert('n', ListNode([2, 3]));
  // node.insert('p', ListNode([3, 4]));
  // node.insert('p', ListNode([4, 5]));
  // console.log(node);
  // const arr = [5, 9, 2, 1, 6];
  // console.log(arr);
  // console.log(quickSort(arr));
}

start();


wip/linked-list/src/list.ts
// type QuoteSymbol = string;
// type BaseSymbol = string;
import { LinkedListSymbol } from './constants';
import { ILinkedList, INode } from './types';

import ListNode from './node';

export default function LinkedList<V>(initial?: Iterable<V>) {
  const list: ILinkedList<V> = {
    [LinkedListSymbol]: true,
    c: 0,
    s: null,
    e: null,
    push(v: V) {
      const node = ListNode(list, v);
      if (!list.e) {
        list.e = node;
        list.s = node;
      } else {
        list.e.insert('n', node);
      }
      return node;
    },
    unshift(v: V) {
      const node = ListNode(list, v);
      if (!list.s) {
        list.s = node;
        list.e = node;
      } else {
        list.s.insert('p', node);
      }
      return node;
    },
    map(cb: (v: V) => any) {
      const arr: any[] = [];
      list.forEach(v => arr.push(cb(v)));
      return arr;
    },
    insert(v: V, cb: (a: V, b: null | V) => any): null | INode<V> {
      let node: null | INode<V> = list.s;
      while (node !== null) {
        const a = node.v;
        const b = node.n ? node.n.v : null;
        if (cb(a, b)) {
          return node.insert('n', v);
        } else if (node === null) {
          break;
        }
        node = node.n;
      }
      return null;
    },
    forEach(cb: (n: V) => void) {
      let node = list.s;
      while (node !== null) {
        cb(node.v);
        node = node.n;
      }
      return list;
    },
    merge(l2: ILinkedList<V>) {
      [...l2].forEach(list.push);
      return list;
    },
    toJSON() {
      return Array.from(list);
    },
    *[Symbol.iterator]() {
      let node = list.s;
      while (node !== null) {
        yield node.v;
        node = node.n;
      }
    },
    reverse: {
      get c() {
        return list.c;
      },
      get s() {
        return list.e;
      },
      get e() {
        return list.s;
      },
      push(v: V) {
        return list.unshift(v);
      },
      unshift(v: V) {
        return list.push(v);
      },
      forEach(cb: (n: V) => void) {
        let node = list.e;
        while (node !== null) {
          cb(node.v);
          node = node.p;
        }
      },
      merge(l2: ILinkedList<V>) {
        [...l2].forEach(list.unshift);
        return list;
      },
      map(cb: (v: V) => any) {
        const arr: any[] = [];
        this.forEach(v => arr.push(cb(v)));
        return arr;
      },
      insert(v: V, cb: (a: V, b: null | V) => any): null | INode<V> {
        let node: null | INode<V> = list.e;
        while (node !== null) {
          const a = node.v;
          const b = node.p ? node.p.v : null;
          if (cb(a, b)) {
            return node.insert('p', v);
          } else if (node === null) {
            break;
          }
          node = node.p;
        }
        return null;
      },
      toJSON() {
        return Array.from(list.reverse);
      },
      *[Symbol.iterator]() {
        let node = list.e;
        while (node !== null) {
          yield node.v;
          node = node.p;
        }
      },
    },
  };
  if (initial) {
    Array.from(initial).forEach(list.push);
  }
  return list;
}

// interface Market<QS extends QuoteSymbol, BS extends BaseSymbol> {
//   readonly market: [QS, BS];

//   orders: OrderBook<QS, BS>;
// }

// interface OrderBook<QS extends QuoteSymbol, BS extends BaseSymbol> {
//   readonly market: Market<QS, BS>;
//   readonly total_asks: string;
//   readonly total_bids: string;

//   readonly precision: number;

//   insert(order: any): boolean;
//   fill(fill: any): boolean;
//   cancel(order: any): boolean;
// }

// const market = {
//   market: ['AURA', 'ETH'],
// };

// const chunk = [
//   // bids
//   [
//     // bid stats
//     [2, 58000000000n, '3'],
//     // bid orders
//     [[30000000000n, '1.5', [25, 26, 27]], [28000000000n, '1.5', [22, 21, 20]]],
//   ],
//   // asks
//   [
//     // ask stats
//     [1, 40000000000n, '0.5'],
//     // ask orders
//     [[40000000000n, '0.5', [28, 29, 30]]],
//   ],
// ];

// console.log(chunk);
